docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+
+     - ``paver watch_assets``
+
+       Paver task that invokes ``webpack --watch`` for Webpack assets and watchdog (a Python library) for other assets.
+
+     - ``assets/build.sh --watch <stage>``
+
+       (where ``<stage>`` is optionally one of the build stages described above. If provided, only that stage's assets will be watched.)
+
+       Bash wrappers around invocation(s) of `watchman <https://facebook.github.io/watchman/>`_, a popular file-watching library maintained by Meta. Watchman is already installed into edx-platform (and other services) via the pywatchman pip wrapper package.
+
+       Note: This adds a Python dependency to build.sh. However, we could be clear that watchman is an *optional* dependency of build.sh which enables the optional ``--watch`` feature. This would keep the *build* action Python-free. Alternatively, watchman is also available Python-free via apt and homebrew.
+
+Migration
+=========
+
+The old asset processing system will be `proposed for deprecation (TODO: link to issue) <http://example.com/depr>`_ upon provisional acceptance of this ADR.
+
+The old and new systems will both be available for at least one named release. Operators will encouraged to try the new asset processing system and report any issues they find. Eventually, the old asset processing system will be entirely removed.
+
+Tutor migration guide
+---------------------
+
+Tutor provides the `openedx-assets <https://github.com/overhangio/tutor/blob/v15.3.0/tutor/templates/build/openedx/bin/openedx-assets>`_ Python script on its edx-platform images for building, collection, and watching. The script uses a mix its own implementation and calls out to edx-platform's paver tasks, avoiding the most troublesome parts of the paver tasks. The script and its interface were the inspiration for the new build.sh that this ADAR describes.
+
+As a consequence of this ADR, Tutor will either need to:
+
+* reimplement the script as a thin wrapper around the new asset processing commands, or
+* deprecate and remove the script.
+
+Either way, the migration path is straightforward:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Existing Tutor-provided command
+     - New upstream command
+   * - ``openedx-assets build``
+     - ``assets/build.sh``
+   * - ``openedx-assets npm``
+     - ``assets/build.sh npm``
+   * - ``openedx-assets xmodule``
+     - ``assets/build.sh xmodule``
+   * - ``openedx-assets common``
+     - ``assets/build.sh common``
+   * - ``openedx-assets themes``
+     - ``assets/build.sh themes``
+   * - ``openedx-assets collect``
+     - ``./manage.py [lms|cms] collectstatic --noinput``
+   * - ``openedx-assets watch-themes``
+     - ``assets/build.sh --watch themes``
+
+The options accepted by ``openedx-assets`` will all be valid inputs to ``assets/build.sh``.
+
+Rejected Alternatives
+*********************
```

Comment:
I expected to see a justification for Bash as a scripting language here. My initial knee-jerk reaction was: why implement this in Bash, and not Python? Of course I'm asking because I'm ~~terrible~~ not so great at Bash, but so are many Open edX contributors. Python makes it much easier to add custom commands, options, etc -- which is why tutor's `openedx-assets` script is in Python.

Is the choice of Bash to make it clear that we are moving away from paver? Or to dissuade other developers from making changes to that script? (which would be a perfectly fine justification IMHO) I think it's great that we are ditching paver, which is yet another build tool where we already have `make` and `npm`. But I'm afraid that we'll end up with a script that is difficult to maintain in the long run.

To be clear: I'm not challenging the choice of Bash vs Python. I'm just curious to better understand the long-term vision.

Reply:
@regisb The driving motivation for Bash are:

### asset build independence 

From the Background:

> Remove all Python dependencies from the static asset build process, such that changes to Python code or requirements do not always have to result in a static asset rebuild.

In other words, the dream is that static assets will be buildable without first copying any Python code or requirement lists from edx-platform into the build context. Only changes to assets or npm requirement lists should require rebuilding the assets.

Now, we could probably still achieve this using Python; however, we would need to ensure that the script remains independent of edx-platform code or requirements. 

### force simplicity

> Or to dissuade other developers from making changes to that script?

Hah, I wouldn't *quite* say that I am trying discourage changes to the script, but I am indeed trying to dissuade folks from expanding its scope or adding layers of indirection to it. 

One of Paver's problems was that it encouraged developers to modularize, build abstracrions, and use clever control flow... because that's what we do when we write Python. But when we're just trying to copy files around & run some shell commands in an expressily NON-extensible way (we aren't trying to add new frontend features to edx-platform!), I find that this tends to increase complexity and reduce readability.

### easier transition to standard tooling

In another comment, you brought up that one of these steps could be an npm post-install step, which I think is awesome. Presuming that the post-install step must be written in JS or Bash, it will be easier to move to that standard tooling now that we have a shell implementation for the npm copying.

In general, I think that switching to standard frontend tools will be easier if our bespoke logic is already distilled down to Bash, rather than tied up in Python logic. In the long term, we hope to delete all of this asset processing junk... if we could delete `build-assets.sh` in favor of `npm build`, that would be a huge win!

Reply:
I will make sure to add those reasons to the ADR. @regisb, do let me know if any of them are unclear or objectionable 🙂

Reply:
I'll incorporate some of that reasoning above into the ADR itself.

Reply:
Done.